Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +26,63 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return totalAmpHoursUsed_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if (current_ < 0)
+       return true;
+    else
+        return false;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    QTime time(0,0,0,0);
+    int ms_remaining;
+      if (isCharging() == true)
+      {
+        ms_remaining = totalAmpHoursUsed() / averageCurrent_ * MS_CONVERSION;
+      }   else
+      {
+        ms_remaining = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed()) / averageCurrent_ * MS_CONVERSION;
+      }
+
+    QTime timeWhenChargedOrDepleted = time.addMSecs(ms_remaining);
+
+    return timeWhenChargedOrDepleted;
+
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
+    int deltaMseconds;
+    double chargeAh;
+    QTime msTime(0,0,0,0);
+
+    if (previousMilliseconds_.isNull())
+    {
+        previousMilliseconds_ = initialTime_.addMSecs(batteryData.time.msec());
+    }
+
+    msTime = msTime.addMSecs(batteryData.time.msec());
```

Comment:
This is basically taking `batteryData.time`, converting it to milliseconds, and then assigning `msTime` to a QTime that is `batteryData.time.msec()` milliseconds ahead. Since the initial value of `msTime` is 0:00:00.000, let's just assign `msTime ` to `batteryData.time`!